In [49]:
import pandas as pd
import numpy as np

## Q1. PANDAS VERSION
What's the version of Pandas that you installed?

'2.3.2'

In [50]:
pd.__version__

'2.3.2'

In [51]:
df = pd.read_csv('car_fuel_efficiency.csv')
df

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                     170            3.0       159.0     3413.433759   
1                     130            5.0        97.0     3149.664934   
2                     170            NaN        78.0     3079.038997   
3                     220            4.0         NaN     2542.392402   
4                     210            1.0       140.0     3460.870990   
...                   ...            ...         ...             ...   
9699                  140            5.0       164.0     2981.107371   
9700                  180            NaN       154.0     2439.525729   
9701                  220            2.0       138.0     2583.471318   
9702                  230            4.0       177.0     2905.527390   
9703                  270            3.0       140.0     2908.043477   

      acceleration  model_year  origin fuel_type         drivetrain  \
0             17.7        2003  Europe  Gasoline    All-wheel drive   
1             17.8        2007     USA  Gasoline  Front-wheel drive   
2             15.1        2018  Europe  Gasoline  Front-wheel drive   
3             20.2        2009     USA    Diesel    All-wheel drive   
4             14.4        2009  Europe  Gasoline    All-wheel drive   
...            ...         ...     ...       ...                ...   
9699          17.3        2013  Europe    Diesel  Front-wheel drive   
9700          15.0        2004     USA  Gasoline    All-wheel drive   
9701          15.1        2008     USA    Diesel    All-wheel drive   
9702          19.4        2011     USA    Diesel  Front-wheel drive   
9703          14.7        2005    Asia    Diesel    All-wheel drive   

      num_doors  fuel_efficiency_mpg  
0           0.0            13.231729  
1           0.0            13.688217  
2           0.0            14.246341  
3           2.0            16.912736  
4           2.0            12.488369  
...         ...                  ...  
9699        NaN            15.101802  
9700        0.0            17.962326  
9701       -1.0            17.186587  
9702        1.0            15.331551  
9703       -1.0            14.884467  

[9704 rows x 11 columns]

## Q2. RECORD QOUNT
How many records are in the dataset?
* 9704

In [52]:
len(df)

9704

## Q3 Fuel types
How many fuel types are presented in the dataset?
* 2

In [53]:
df['fuel_type'].nunique()

2

## Q4 MISSING VALUES
How many columns in the dataset have missing values?
* 4 (num_cylinders, horsepower, acceleration, num_doors)

In [54]:
df.isnull().sum()

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

## Q5. Max fuel efficiency
What's the maximum fuel efficiency of cars from Asia?
* 23.75

In [55]:
df[df['origin'] == 'Asia']['fuel_efficiency_mpg'].max()

23.759122836520497

## Q6 Median value of horsepower
1. Find the median value of `horsepower` column in the dataset.
2. Next, calculate the most frequent value of the same `horsepower` column.
3. Use `fillna` method to fill the missing values in `horsepower` column with the most frequent value from the previous step.
4. Now, calculate the median value of `horsepower` once again.

Has it changed?
- Yes, it increased

In [56]:
#find the median value of horsepower column in the dataset.
df['horsepower'].median()

149.0

In [57]:
# 2. Most frequent value (mode)
most_frequent = df['horsepower'].mode()[0]
most_frequent 

np.float64(152.0)

In [58]:
# 3. Fill missing values with the most frequent value
df['horsepower'] = df['horsepower'].fillna(most_frequent)

In [59]:
# 4. Median value after filling missing values
median_after = df['horsepower'].median()
median_after

152.0

## Q7. Sum of weights

1. Select all the cars from Asia
2. Select only columns `vehicle_weight` and `model_year`
3. Select the first 7 values
4. Get the underlying NumPy array. Let's call it `X`.
5. Compute matrix-matrix multiplication between the transpose of `X` and `X`. To get the transpose, use `X.T`. Let's call the result `XTX`.
6. Invert `XTX`.
7. Create an array `y` with values `[1100, 1300, 800, 900, 1000, 1100, 1200]`.
8. Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.
9. What's the sum of all the elements of the result?

* 0.51

In [60]:
# 1-3. Select cars from Asia, columns, and first 7 rows
X = df[df['origin'] == 'Asia'][['vehicle_weight', 'model_year']].head(7).to_numpy()

In [61]:
# 4-5. Compute XTX
XTX = X.T @ X

In [62]:
# 6. Invert XTX
XTX_inv = np.linalg.inv(XTX)

In [63]:
# 7. Create y
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])

In [64]:
# 8. Compute w
w = XTX_inv @ X.T @ y

In [65]:
# 9. Sum of all elements in w
w.sum()

np.float64(0.5187709081074007)